grakn-graph/src/main/java/ai/grakn/graph/internal/TypeImpl.java

```diff
@@ -95,8 +104,16 @@ protected V addInstance(Schema.BaseType instanceBaseType, BiFunction<Vertex, T,
     @Override
     public Collection<RoleType> playsRoles() {
         Set<RoleType> allRoleTypes = new HashSet<>();
-        getSuperSet().forEach(type -> allRoleTypes.addAll(((TypeImpl <?, ?>) type).getOutgoingNeighbours(Schema.EdgeLabel.PLAYS_ROLE)));
-        return filterImplicitStructures(allRoleTypes);
+
+        //Get the immediate plays roles which may be cached
+        allRoleTypes.addAll(cachedPlaysRoles.get());
+
+        //Now get the super type plays roles (Which may also be cached locally within their own context
+        Set<T> superSet = getSuperSet();
+        superSet.remove(this); //We already have the plays roles from ourselves
+        superSet.forEach(superParent -> allRoleTypes.addAll(superParent.playsRoles()));
```

Comment:
I think you have a combinatorial explosion here. For example, if you have `a -subs-> b -subs-> c -subs-> d`, then `a.playsRoles()` will call `b.playsRoles()`, `c.playsRoles()` and `d.playsRoles()`.
`b.playsRoles()` will then call both `c.playsRoles()` and `d.playsRoles()` etc.

You can deal with this by either only calling something like `b.directPlaysRoles()`, `c.directPlaysRoles()`. Or, you just call on the direct super-type only: `b.playsRoles()`